File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/MemberExpression.java
package me.itzisonn_.meazy.parser.ast.expression;

import lombok.Getter;
import lombok.Setter;
import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.parser.ast.statement.LocalStatement;
import me.itzisonn_.meazy.runtime.environment.Environment;
import org.jspecify.annotations.NullMarked;

@Getter
@NullMarked
public class MemberExpression implements Expression, LocalStatement {
    @Setter
    private Expression object;
    private final Expression member;
    private final boolean isNullSafe;

    public MemberExpression(Expression object, Expression member, boolean isNullSafe) {
        this.object = object;
        this.member = member;
        this.isNullSafe = isNullSafe;
    }

    @Override
    public void emit(InstructionsSet instructions, Environment environment, ProgramUnit parent) {
        member.emit(instructions, environment, this);
    }

    @Override
    public DataType getType(Environment environment, ProgramUnit parent) {
        return member.getType(environment, this);
    }

    @Override
    public boolean alwaysReturns() {
        return false;
    }

}

File: src/main/kotlin/me/itzisonn_/meazy/parser/ast/expression/MemberExpression.kt
package me.itzisonn_.meazy.parser.ast.expression

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.parser.ast.ParentMap
import me.itzisonn_.meazy.runtime.data.DataType
import me.itzisonn_.meazy.parser.ast.statement.LocalStatement
import me.itzisonn_.meazy.runtime.environment.Environment

class MemberExpression(
    val receiver: Expression,
    val member: Expression,
    val isNullSafe: Boolean
) : Expression, LocalStatement {
    override val children = setOf(receiver, member)

    context(parents: ParentMap)
    override fun emit(instructions: InstructionsSet, environment: Environment) {
        member.emit(instructions, environment)
    }

    context(parents: ParentMap)
    override fun getType(environment: Environment): DataType {
        return member.getType(environment)
    }

    override fun alwaysReturns() = false
}